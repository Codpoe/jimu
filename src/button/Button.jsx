import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Loader from '../loader/';
import './button.css';

export default class Button extends Component {
    constructor(props) {
        super(props);
        // a
        this.handleClick = this.handleClick.bind(this);
    }

    handleClick(ev) {
        const { onClick } = this.props;
        onClick && onClick(ev);
    }

    render() {
        const {
            type,
            size,
            text,
            outline,
            flat,
            round,
            circle,
            block,
            disabled,
            loading,
            href,
            target,
            loaderColor,
            children,
            className,
            style
        } = this.props;

        const Node = href ? 'a' : 'button';

        const btnClass = classnames(className, 'my-button', {
            [`my-button--${type}`]: true,
            [`my-button--${size}`]: true,
            'my-button--text': text,
            'my-button--outline': outline,
            'my-button--flat': flat,
            'my-button--round': round,
            'my-button--circle': circle,
            'my-button--block': block,
            'my-button--disabled': disabled,
            'my-button--loading': loading
        });

        return (
            <Node
                className={btnClass}
                style={style}
                href={href}
                target={target}
                disabled={disabled}
                onClick={this.handleClick}
            >
                {children}
                {loading && (
                    <Loader color={loaderColor}/>
                )}
            </Node>
        )
    }
}

Button.propTypes = {
    type: PropTypes.oneOf(['default', 'primary', 'success', 'warning', 'error']),
    size: PropTypes.oneOf(['normal', 'large', 'small']),
    text: PropTypes.bool,
    outline: PropTypes.bool,
    flat: PropTypes.bool,
    round: PropTypes.bool,
    circle: PropTypes.bool,
    block: PropTypes.bool,
    disabled: PropTypes.bool,
    loading: PropTypes.bool,
    onClick: PropTypes.func,
    loaderColor: PropTypes.string,
    className: PropTypes.string,
    style: PropTypes.object
}

Button.defaultProps = {
    type: 'default',
    size: 'normal',
    text: false,
    outline: false,
    flat: false,
    round: false,
    circle: false,
    block: false,
    disabled: false,
    loading: false,
    loaderColor: 'white',
    className: '',
    style: {}
}