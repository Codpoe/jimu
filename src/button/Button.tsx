import React from 'react';
import cx from 'classnames';
import bem from '../utils/bem';
import Loader from '../loader/index';
import './style/button.css';

export interface ButtonProps {
  type?: 'default' | 'primary' | 'success' | 'warning' | 'error';
  size?: 'normal' | 'large' | 'small';
  hollow?: boolean;
  text?: boolean;
  pure?: boolean;
  round?: boolean;
  fullRound?: boolean;
  block?: boolean;
  disabled?: boolean;
  loading?: boolean;
  href?: string;
  target?: '_self' | '_blank';
  loaderColor?: string;
  onClick?: React.EventHandler<any>;
  className?: string;
  style?: React.CSSProperties;
}

const b = bem('jm-button');
export default class Button extends React.Component<ButtonProps> {
  static defaultProps = {
    type: 'default',
    size: 'normal',
    hollow: false,
    text: false,
    pure: false,
    round: false,
    fullRound: false,
    block: false,
    disabled: false,
    loading: false,
    loaderColor: 'white',
    href: '',
    target: '_blank',
    onClick: () => {},
  };

  render() {
    const {
      type,
      size,
      hollow,
      text,
      pure,
      round,
      fullRound,
      block,
      disabled,
      loading,
      href,
      target,
      loaderColor,
      children,
      onClick,
      className,
      style,
    } = this.props;

    const NodeName = href ? 'a' : 'button';

    const cls = cx(className, b(), b([type]), b([size]), {
      [b(['hollow'])]: hollow,
      [b(['text'])]: text || pure,
      [b(['pure'])]: pure,
      [b(['round'])]: round,
      [b(['full-round'])]: fullRound,
      [b(['block'])]: block,
      [b(['disabled'])]: disabled,
      [b(['loading'])]: loading,
    });

    return (
      <NodeName
        className={cls}
        style={style}
        href={href}
        target={href ? target : undefined}
        disabled={disabled}
        onClick={onClick}
      >
        {loading ? <Loader color={loaderColor} /> : children}
      </NodeName>
    );
  }
}
