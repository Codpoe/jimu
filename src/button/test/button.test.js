import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Button from '../button/';

describe('Button test', () => {
  it('snapshot', () => {
    const button = renderer.create(<Button>TEST</Button>);
    const tree = button.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('default', () => {
    const button = shallow(<Button>TEST</Button>);
    expect(button.type()).toBe('button');
    expect(button.hasClass('rdf-button')).toBeTruthy();
    expect(button.hasClass('rdf-button--default')).toBeTruthy();
    expect(button.childAt(0).text()).toBe('TEST');
  });

  it('type', () => {
    const jm = shallow(<Button type="primary">TEST</Button>);
    expect(jm.hasClass('rdf-button--primary')).toBeTruthy();
  });

  it('size', () => {
    const jm = shallow(<Button size="large">TEST</Button>);
    expect(jm.hasClass('rdf-button--large')).toBeTruthy();
  });

  it('text', () => {
    const jm = shallow(<Button text>TEST</Button>);
    expect(jm.hasClass('rdf-button--text')).toBeTruthy();
  });

  it('outline', () => {
    const jm = shallow(<Button outline>TEST</Button>);
    expect(jm.hasClass('rdf-button--outline')).toBeTruthy();
  });

  it('flat', () => {
    const jm = shallow(<Button flat>TEST</Button>);
    expect(jm.hasClass('rdf-button--flat')).toBeTruthy();
  });

  it('round', () => {
    const jm = shallow(<Button round>TEST</Button>);
    expect(jm.hasClass('rdf-button--round')).toBeTruthy();
  });

  it('block', () => {
    const jm = shallow(<Button block>TEST</Button>);
    expect(jm.hasClass('rdf-button--block')).toBeTruthy();
  });

  it('disabled', () => {
    const jm = shallow(<Button disabled>TEST</Button>);
    expect(jm.hasClass('rdf-button--disabled')).toBeTruthy();
  });

  it('loading', () => {
    const jm = shallow(<Button loading>TEST</Button>);
    expect(jm.hasClass('rdf-button--loading')).toBeTruthy();
  });

  it('a', () => {
    const jm = shallow(<Button href="https://youzan.com">TEST</Button>);
    expect(jm.type()).toBe('a');
  });

  it('click', () => {
    const fn = jest.fn();
    const jm = shallow(<Button onClick={fn}>TEST</Button>);
    jm.simulate('click');
    expect(fn.mock.calls.length).toBe(1);
  });
});
