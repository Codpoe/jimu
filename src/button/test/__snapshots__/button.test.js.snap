// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button test snapshot 1`] = `
<button
  className="rdf-button rdf-button--default rdf-button--normal"
  disabled={false}
>
  TEST
</button>
`;
